### PR TITLE
Expose GrpcClient to bridge layer via factory functions

### LINKED ISSUE #12: Connect Client factory to bridge layer
Blocked by #11.

- [x] Claim an `ex_data_index` in `InitEngine` on the `ENGINE` struct (for a pointer to a `GrpcApiClient` instance)
- [x] Instantiate `GrpcApiClient` in `InitEngine`
- [x] Add to `ENGINE` via `ENGINE_set_ex_data`
- [x] Free `GrpcApiClient` in `FinishEngine`
- [x] Write unit tests (using `ENGINE_get_ex_data`) to make sure `GrpcApiClient` was instantiated

### PR DESCRIPTION
**Pull request by ZacharyEspiritu**
Resolves #12.

### CODE REVIEW

File: src/backing/client/client_factory.cc [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+using ::kmsengine::backing::grpc_client::GrpcClient;
+
+// Default API endpoint for making gRPC requests against.
+//
+// Should not be prefixed with "https://" to comply with the gRPC library.
+constexpr char kDefaultApiEndpoint[] = "cloudkms.googleapis.com";
+
+std::unique_ptr<Client> MakeDefaultClient(
+    absl::optional<std::chrono::milliseconds> timeout) {
+  auto credentials = grpc::GoogleDefaultCredentials();
```
Comment (ajfeldma-google) [inline src/backing/client/client_factory.cc]: Nit: remove unnecessary variables?

File: src/backing/client/client_factory_test.cc [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace kmsengine {
+namespace backing {
+namespace {
+
+TEST(ClientFactoryTest, MakeDefaultClientWithoutTimeout) {
+  // We can't make calls with an authenticated client here, so for now just
+  // check that the factory function is callable and returns a success
+  // `Status`.
+  (void)MakeDefaultClientWithoutTimeout();
```
Comment (ajfeldma-google) [inline src/backing/client/client_factory_test.cc]: These functions don't return `Status`, right? What happens on error?

**Review by ajfeldma-google — CHANGES_REQUESTED** [review]

File: src/backing/client/client_factory_test.cc (hunk 2 above)
Comment (ZacharyEspiritu) [reply src/backing/client/client_factory_test.cc]: It seems that in `grpc` the authentication isn't checked until an actual call is made. At the very least, I can't find any instance of anyone error-checking the result from `grpc::GoogleDefaultCredentials`. (gRPC will return an error status on an RPC call with bad credentials.)

Comment (ZacharyEspiritu) [reply src/backing/client/client_factory_test.cc]: The call to `grpc::GoogleDefaultCredentials` should be the only thing that can "fail" here, though as stated above, the "failure" isn't clear until later.

File: src/backing/client/client_factory.cc [hunk 3]
```diff
@@ -41,12 +41,11 @@ constexpr char kDefaultApiEndpoint[] = "cloudkms.googleapis.com";
 
 std::unique_ptr<Client> MakeDefaultClient(
     absl::optional<std::chrono::milliseconds> timeout) {
-  auto credentials = grpc::GoogleDefaultCredentials();
-  auto stub = grpc_client::CreateKeyManagementServiceStub(kDefaultApiEndpoint,
-                                                          credentials);
-
-  auto clock = std::make_shared<client::SystemClock>();
-  auto factory = grpc_client::CreateClientContextFactory(timeout, clock);
+  auto stub = grpc_client::CreateKeyManagementServiceStub(
```
Comment (ajfeldma-google) [inline src/backing/client/client_factory.cc]: I think if you just move these into `absl::make_unique<GrpcClient>(…)` then you don't even need `std::move`

**Review by ajfeldma-google — CHANGES_REQUESTED** [review]

**Review by ajfeldma-google — DISMISSED** [review]

File: src/backing/client/client_factory_test.cc [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+TEST(ClientFactoryTest, MakeDefaultClientWithTimeout) {
+  // We can't make calls with an authenticated client here, so for now just
+  // check that the factory function is callable and returns a success
+  // `Status`.
+  (void)MakeDefaultClientWithTimeout(std::chrono::milliseconds(50));
```
Comment (tonyliaoss) [inline src/backing/client/client_factory_test.cc]: Is the `void` here to get rid of some compiler warnings about not using the result?

Maybe do something like `EXPECT_THAT(MakeDefaultClientWithoutTimeout(), NotNull())` or something?

**Review by ajfeldma-google — APPROVED** [review]

**Review by tonyliaoss — APPROVED** [review]

**Pull request merged**